src/connector/src/sink/file_sink/batching_log_sink.rs

```diff
@@ -0,0 +1,141 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use async_trait::async_trait;
+
+use crate::sink::file_sink::opendal_sink::OpenDalSinkWriter;
+use crate::sink::log_store::{LogStoreReadItem, TruncateOffset};
+use crate::sink::{LogSinker, Result, SinkLogReader};
+
+/// `BatchingLogSinker` is used for a commit-decoupled sink that supports cross-barrier batching.
+/// Currently, it is only used for file sinks, so it contains an `OpenDalSinkWriter`.
+/// When the file sink writer completes writing a file, it truncates the corresponding `chunk_id` in the log store.
+pub struct BatchingLogSinker {
+    writer: OpenDalSinkWriter,
+}
+
+impl BatchingLogSinker {
+    /// Create a log sinker with a file sink writer.
+    pub fn new(writer: OpenDalSinkWriter) -> Self {
+        BatchingLogSinker { writer }
+    }
+}
+
+#[async_trait]
+impl LogSinker for BatchingLogSinker {
+    async fn consume_log_and_sink(self, log_reader: &mut impl SinkLogReader) -> Result<!> {
+        let mut sink_writer = self.writer;
+        #[derive(Debug)]
+        enum LogConsumerState {
+            /// Mark that the log consumer is not initialized yet
+            Uninitialized,
+
+            /// Mark that a new epoch has begun.
+            EpochBegun { curr_epoch: u64 },
+
+            /// Mark that the consumer has just received a barrier
+            BarrierReceived { prev_epoch: u64 },
+        }
+
+        let mut state = LogConsumerState::Uninitialized;
+        loop {
+            let (epoch, item): (u64, LogStoreReadItem) = log_reader.next_item().await?;
+            if let LogStoreReadItem::UpdateVnodeBitmap(_) = &item {
+                match &state {
+                    LogConsumerState::BarrierReceived { .. } => {
+                        // we need to force to finish the batch here. Otherwise, there can be data loss because actor can be dropped and rebuilt during scaling.
+                        if let Some(committed_chunk_id) = sink_writer.should_finish().await? {
+                            log_reader.truncate(TruncateOffset::Chunk {
```

Comment:
Should we use `TruncateOffset::Barrier` here? cc @wenym1

Reply:
Just checked the two log store implementations, they both have logic assuming that we call `truncate` on `TruncateOffset::Barrier`. So we'd better ensure that we call a `truncate` on barrier.

Reply:
Do we have to call `truncate` on every barrier? Will there be any issue if we skip some of them?

Reply:
I think we should do 
```rust
if should_finish(){
   truncate(barrier);
   truncate(chunk);
}else{
   skip epoch, do nothing
}
```

Reply:
Talked with @wenym1 : currently, if `sink_decouple=false`, they need to be truncated one by one; if `sink_decouple=true`, the barrier can be skipped.
Sink this pr ban `sink_decouple = false` for file sink, so I think 
```
if should_finish(){
      1. truncate(barrier); 
      2. truncate(chunk); 
}
else{ 
       skip epoch, do nothing
 }
```
is fine.

Let's change the behavior of log reader in another pr to let them only  truncate the newest barrier.

Reply:
The ordering here is wrong. Logically you can view TruncateOffset::Barrier(epoch) as equivalent to TruncateOffset::Chunk(epoch, u64::max) (See [here](https://github.com/risingwavelabs/risingwave/blob/c6e616b3606f52cad14b94b3847e159347c78ee6/src/connector/src/sink/log_store.rs#L41)) so the codes here makes the truncate offset rewind. Instead, you only need to call TruncateOffset::Barrier(epoch) without TruncateOffset::Chunk here.